Build Script Improvements and Update Automation

ProjectSynchro opened the pull request:
Updates the application to version 0.8.0, and the FreeDesktop runtime to 24.08.

The manifest has been updated to consume upstream deb packages, enable support for Wayland and uses the external data checker for updates.
I also added an updated metainfo which includes new version information, as well as other information related to the project.

This changes this flatpak to be built from existing binaries. This is mainly due to the difficulty/feasibility behind building using electron-forge offline as well as the lack of support for v3 npm lockfiles in [flatpak-builder-tools](https://github.com/flatpak/flatpak-builder-tools). 

Currently this isn't possible with my testing (at least with the current flatpak tools) but if someone figures this out going back to source building is the way to go. Due to the fact that both AMD64 and ARM64 debian packages are available upstream, nobody loses out on access to this application in the meantime however.

Let me know if you'd like any changes.

Obsoletes #7

hadess commented:
> This changes this flatpak to be built from existing binaries. This is mainly due to the difficulty/feasibility behind building using electron-forge offline as well as the lack of support for v3 npm lockfiles in [flatpak-builder-tools](https://github.com/flatpak/flatpak-builder-tools).

I'll let @FelicianoTech figure out whether this is acceptable. In my opinion, this should be worked on (or at least attempted to be worked on) upstream so that the package keeps being built from source.

I should also note that the commits don't seem to be organised correctly, the package should still be building after each commit...

hadess commented on local-build.sh:
```diff
@@ -1,10 +0,0 @@
-#!/usr/bin/env bash
-
-BRANCH=test
```
Why was this deleted?

ProjectSynchro replied:
I figured the instructions in the README were sufficient, but I added it back as there isn't really a reason to remove it.

ProjectSynchro replied:
Solved with https://github.com/flathub/tech.feliciano.pocket-casts/pull/8/commits/24f38ad1eca48114a9492ca12e7dc3e39d2198bf

hadess commented on 0001-flatpak-Add-missing-item-in-appdata-file.patch:
```diff
@@ -1,31 +0,0 @@
-From 3d7d7dad71e7bd89bd00c0aa99767c0438c3beeb Mon Sep 17 00:00:00 2001
```
You're removing files before they're obsolete. This would break the build.

hadess commented on tech.feliciano.pocket-casts.metainfo.xml:
```diff
@@ -0,0 +1,80 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This file should be upstream. You removed the appdata file in the previous commit, those are the same files. Please make sure to keep this being a patch if the upstream file still exists.

ProjectSynchro replied:
Metainfo is not provided in the upstream package releases unfortunately. I could patch the git repository with these changes instead, I already use that source for the upstream icon. 

The package that is shipped doesn't provide the right icon either.

hadess replied:
Right, those could be things that get fixed upstream as well, or at least filed as issues upstream.

In the case of the metainfo file, it would be better to fetch the file from upstream, and apply a small patch to it locally, rather than ship a whole modified file, as that would make it harder to maintain.

ProjectSynchro replied:
Indeed, I'll open some a PR upstream shortly, the desktop file and metainfo could use some sprucing up.

hadess commented on tech.feliciano.pocket-casts.yml:
```diff
@@ -1,67 +1,99 @@
+            ]]; then
+          - FLAGS="$FLAGS --enable-features=WaylandWindowDecorations --ozone-platform-hint=auto"
+          - fi
+          - FLATPAK_HOST=1 zypak-wrapper.sh /app/lib/youtube-music-desktop-app/youtube-music-desktop-app
```
`youtube-music-desktop-app` ?

ProjectSynchro replied:
Bad commit, wasn't aware it pushed.

ProjectSynchro commented:
> I should also note that the commits don't seem to be organised correctly, the package should still be building after each commit...

Seems like my changes didn't get correctly applied, force pushed so they are in the correct order, as well as removed some bad commits.

ProjectSynchro commented:
I'm going to mark this as a draft, I may be able to get this to build from source after all.

ProjectSynchro commented:
bot, build

ProjectSynchro commented:
Okay, round two.

I have figured out how to get source builds working again, and have bumped every dependency I could to the newest that is supported (FreeDesktop, Node).

Additionally I have a solution for managing automatic updates. 

Normally if you aren't building a node app you can use [flatpak-external-data-checker](https://github.com/flathub-infra/flatpak-external-data-checker) and it will bump the manifest and metainfo automatically (the metainfo is handy if there isn't anything upstream, mainly just adds a placeholder version tag though)

With node however we need to generate the NPM cache so the build works while offline. I implemented a Github Actions workflow that will update the generated-sources.json alongside any external-data-checker checks. 

The external data checker will then create a PR for the update (if there is one) which gets test built by CI and then can be merged by maintainers if it works correctly.

Let me know if you'd like any changes. I'm pretty sure the commits can be built cleanly as you had mentioned you'd like it to be that way.

ProjectSynchro commented:
Some additional context on the Github Actions workflow:

It uses the docker based action provided by flathub, so external dependency-wise that is the only one, actions/checkout is maintained by Github and doesn't pose any issues in terms of maintainability.

Essentially the workflow checks with the external data checker if there is an update in an edit only mode, I then use git to check if there is any changes to files. 

If there is, we get the latest upstream tagged release's `package.json` and `package-lock.json` and run `flatpak-node-generator` to get our `generated-sources.json`. 

From there the git repository has the manifest file reset and the external data checker is called again in merge mode (which opens a PR with changes) that will open the PR with the manifest updates and the `generated-sources.json` updates.

Hopefully that helps outline what the workflow does.

ProjectSynchro commented:
Still TODO, open a PR upstream with the metainfo and desktop file changes so we can get rid of the patches eventually.

ProjectSynchro commented:
bot, build

ProjectSynchro commented:
Fixed a stray command. Build [149722](https://buildbot.flathub.org/#/builders/6/builds/149722) will fail but the next one should be good now.

hadess commented on local-build.sh:
```diff
@@ -1,10 +1,10 @@
 rm -f tech.feliciano.pocket-casts.flatpak
-rm -rf _build ; mkdir _build
-rm -rf _repo ; mkdir _repo
+rm -rf build-dir ; mkdir build-dir
```
I don't think those changes are necessary.

ProjectSynchro replied:
I tend to use `build-dir` myself but I can revert these changes.

hadess commented on tech.feliciano.pocket-casts.yml:
```diff
@@ -1,67 +1,83 @@
+command: start-pocket-casts
 separate-locales: false
 finish-args:
+  # Share IPC namespace with the host, without it the X11 shared memory extension will not work.
```
Can you make the changes related to documenting the finish args as a separate commit? This would make the runtime update specific changes clearer.

ProjectSynchro replied:
Sure, I can do that.

hadess commented on 001-update-metainfo.patch:
```diff
@@ -0,0 +1,98 @@
+   <name>Pocket Casts</name>
++  <summary>Linux desktop player for Pocket Casts</summary>
+   <developer_name>Ricardo N Feliciano</developer_name>
+-  <summary>Online podcast player</summary>
```
Why the summary change?

ProjectSynchro replied:
I initially was thinking I would change it to be more descriptive. Looking now: https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/quality-guidelines#summary might not want to do that.

It may also be worth it to copy what is used for the snap for consistency's sake: https://github.com/FelicianoTech/pocket-casts-desktop-app/blob/06d0439647b500b0ab84566d2d9028d1cff5ede6/package.json#L45

Main issue with that is it may not work out to be totally within the quality guidelines.

hadess commented on 001-update-metainfo.patch:
```diff
@@ -0,0 +1,98 @@
+     <screenshot type="default">
+-      <image type="source">https://raw.githubusercontent.com/felicianotech/pocket-casts-desktop-app/trunk/screenshot.png</image>
++      <caption>A screenshot of the main interface</caption>
++      <image>https://raw.githubusercontent.com/felicianotech/pocket-casts-desktop-app/v0.8.0/screenshot.png</image>
```
This screenshot isn't suitable, please revert.

Those are the guidelines in case you're interested in updating the screenshot:
https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/quality-guidelines#screenshots

ProjectSynchro replied:
I see, I had just continued to use the upstream metainfo's screenshot. 

Didn't really want to change it but I'll take some screenshots and stick them in the repo here instead.

hadess commented on 001-update-metainfo.patch:
```diff
@@ -0,0 +1,98 @@
+   </screenshots>
+-  <content_rating type="oars-1.1" />
+   <releases>
+-    <release version="0.5.0" date="2020-12-13">
```
Why remove this metadata?

ProjectSynchro replied:
Mainly for consistency and maintenance sake. (there were some formatting changes too so those lines were removed but re-added later in the diff. 

It's far simpler to just point to the github release for each of the releases. The upstream developer could change the information to include additional information, or in the case of v0.8.0 the release notes just point to additional links to github PRs. 

Some past releases don't have any changelogs at all.

hadess commented on README.md:
```diff
@@ -1,7 +1,22 @@
+### Build the Flatpak locally
+1. Clone this repo
+2. Install the flatpak-builder package for your distro
+3. Install FreeDesktop SDK and Runtime, as well as the Electron BaseApp
```
Maybe making the script check for missing runtimes would be better?

I've done this in https://github.com/flathub/org.ghidra_sre.Ghidra/blob/master/generate-deps.sh#L15

So you could remove step 3 and have step 4 say something like:
If necessary SDKs are missing, the script will error out and prompt you to install those.

ProjectSynchro replied:
I was thinking about doing that so the workflow can reference it too. That should be pretty doable.

ProjectSynchro replied:
Implemented this in local-build.sh

https://github.com/ProjectSynchro/tech.feliciano.pocket-casts/commit/e53d0a71852fdb9ee658a1d8ad7d40906b07996c

hadess commented on README.md:
```diff
@@ -1,7 +1,22 @@
+   * `git submodule update --init`
+3. Install flatpak-node-generator with:
+   * `pipx install flatpak-builder-tools/node`
+4. Download the latest package-lock.json from the upstream repository
```
Maybe you could add a copy/paste command for that too? Using curl or wget would be fine.

ProjectSynchro replied:
Sure, seems simple enough :)

ProjectSynchro replied:
I implemented this in a script (generate_deps.sh)
https://github.com/ProjectSynchro/tech.feliciano.pocket-casts/commit/055a8924aded5588be28131962bc9e227fd96f7a

hadess reviewed (changes requested):
I have no idea about the implications of using GitHub actions, it's something that "somebody" (you or I, or anybody else that might be reading this) would need to research.

I'm happy to commit the rest of the commits after the review changes are applied, if you want to put the GitHub actions changes in a separate PR.

ProjectSynchro commented:
> I have no idea about the implications of using GitHub actions, it's something that "somebody" (you or I, or anybody else that might be reading this) would need to research.
> 
> I'm happy to commit the rest of the commits after the review changes are applied, if you want to put the GitHub actions changes in a separate PR.

I can do that. I have tested it on my fork already and it seems to work well.

See here for a PR example: https://github.com/ProjectSynchro/tech.feliciano.pocket-casts/pull/3

And See the worflow run here: https://github.com/ProjectSynchro/tech.feliciano.pocket-casts/actions/runs/11058580552/job/30724922447

hadess commented:
I've reordered and merged most of the necessary changes for the build in https://github.com/flathub/tech.feliciano.pocket-casts/pull/10

You can remove `Bump FreeDesktop Runtime to 24.08, bump node to node20` from this and keep only the dependency automation to avoid opening a new PR.

The appdata and desktop file changes will definitely have to go through upstream, rather than the very large changes in this PR.

ProjectSynchro commented:
Apologies, got swamped with work.

I'll clean this up to just include the workflow commits and changes.

ProjectSynchro committed:
Add .gitignore

This stops build directories and files from accidentally being staged.

ProjectSynchro committed:
flathub.json: Disable external data checker

This prevents any updates to the manifest via the external data checker due to the need to update dependencies externally.

ProjectSynchro committed:
local-build.sh: Quote variables to prevent potential problems

ProjectSynchro commented:
Not quite sure why Github closed this, oh well.

I'll set it as a draft for now.

ProjectSynchro committed:
local-build.sh: Add checks for dependencies before building

This should make this a one shot script, assuming flatpak and flatpak-builder are installed.

ProjectSynchro committed:
Add flatpak-builder-tools as a submodule

ProjectSynchro committed:
Add generate_deps.sh script

This will check for dependencies and generate an updated generated-sources.json.

ProjectSynchro committed:
Update README

ProjectSynchro committed:
Add Github Actions workflow for updating dependencies alongside application updates.

This uses flatpak-external-data-checker itself to update the manifest, as well as node dependencies using flatpak-node-generator.

ProjectSynchro commented:
Ready to go now.

hadess commented on local-build.sh:
```diff
@@ -1,10 +1,70 @@
-rm -rf _build ; mkdir _build
-rm -rf _repo ; mkdir _repo
+rm -rf _build _repo
+mkdir _build _repo
```
The directories will be re-created, that can be removed.

hadess replied:
That review is still relevant, you can remove the `mkdir` calls.

ProjectSynchro replied:
Ah I can open a PR to fix this if you want this adjusted, must have skipped my mind.

hadess commented:
> I have no idea about the implications of using GitHub actions, it's something that "somebody" (you or I, or anybody else that might be reading this) would need to research.

Discussion is here:
https://discourse.flathub.org/t/using-github-actions-workflow-to-update-deps/7816

hadess commented:
I've rebased this, but it still needs changes as the app has slightly changed, and we need to apply some manual changes on top of the current codebase to make it build. The dependency generator is also still broken upstream.

See https://github.com/flathub/tech.feliciano.pocket-casts/pull/15

ProjectSynchro commented:
I'll take a look at this again and see if I can get it mergable.

ProjectSynchro commented:
Thanks to @Ian2020's implementation in https://github.com/flatpak/flatpak-builder-tools/pull/382 pulling updates for the node dependencies that use git sources is trivial. 

I've updated the git submodule in use to track the gitrefs branch in their fork. When that PR is upstreamed this can be changed to the upstream https://github.com/flatpak/flatpak-builder-tools repo.

I ran a test build with the generated sources outputted by the script and all seems well as far as test building.

ProjectSynchro committed:
Change submodule to point to Ian2020's branch of flatpak-builder-tools

This is due to https://github.com/flatpak/flatpak-builder-tools/issues/381

Their PR https://github.com/flatpak/flatpak-builder-tools/pull/382 implements the required functionality.

ProjectSynchro commented:
This is ready to go.

Here is a sample PR for an update from 0.10.1 to 0.10.2: https://github.com/ProjectSynchro/tech.feliciano.pocket-casts/pull/4

The generated commit is here: https://github.com/ProjectSynchro/tech.feliciano.pocket-casts/pull/4/commits/c04f4c22160766a4129829995a1002ae7920e3ed
